src/main/java/net/neoforged/neoforge/capabilities/IBlockCapabilityProvider.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.capabilities;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.state.BlockState;
+import org.jetbrains.annotations.Nullable;
+
+@FunctionalInterface
+public interface IBlockCapabilityProvider<T, C> {
+    /**
+     * Returns the capability, or {@code null} if not available.
+     * <p>
+     * <b>If a previously returned capability is not valid anymore, or if a new capability is available,
+     * {@link Level#invalidateCapabilities(BlockPos)} MUST be called to notify the caches (see below).</b>
+     * <p>
+     * Capabilities are automatically invalidated by Forge in the following cases:
+     * <ul>
+     * <li>Chunk loads and unloads.</li>
+     * <li>Block entity loads and unloads.</li>
+     * <li>Block entity placement and destruction.</li>
+     * </ul>
+     * In ALL other cases, it is the responsibility of the modder to call {@link Level#invalidateCapabilities(BlockPos)},
+     * for example if the configuration of a block entity changes.
+     *
+     * @param level       The level.
+     * @param pos         The position.
+     * @param state       The block state.
+     * @param blockEntity The block entity, if present. {@code null} means that there is no block entity at the target position.
+     * @param context     Extra context, capability-dependent.
+     */
+    // TODO: Level, or a restricted version like BlockGetter/LevelAccess?
+    @Nullable
+    T getCapability(Level level, BlockPos pos, BlockState state, @Nullable BlockEntity blockEntity, C context);
```

Comment:
This is quite a tricky question because many things are not possible on a `LevelAccessor` - you can't invalidate caps with a `LevelAccessor` at the moment, for example.

Reply:
Is there anything that prevents us from exposing those methods from LevelAccessor?

Reply:
We would have to track the lifecycle of `WorldGenRegion`s, which I'm not sure is a good idea.

I also noticed that `BlockEntity#level` is of the `Level` type, and therefore not set in `WorldGenRegion`s... Just this might be a good reason to stick with `Level`.

Reply:
I mean getCapability is nullable anyway so we could just use a noop one in WorldGenRegion

Reply:
Let's stick with `Level`